lib/spack/spack/ci/__init__.py

```diff
@@ -0,0 +1,1296 @@
+            "Failed to automatically setup the tested version of spack "
+            "in your local reproduction directory."
+        )
+        print(setup_msg)
+
+    # In cases where CI build was run on a shell runner, it might be useful
+    # to see what tags were applied to the job so the user knows what shell
+    # runner was used.  But in that case in general, we cannot do nearly as
+    # much to set up the reproducer.
+    job_tags = None
+    if "tags" in job_yaml:
+        job_tags = job_yaml["tags"]
+        tty.msg(f"Job ran with the following tags: {job_tags}")
+
+    entrypoint_script = [
+        ["git", "config", "--global", "--add", "safe.directory", mount_as_dir],
+        [
+            ".",
+            os.path.join(
+                mount_as_dir if job_image else work_dir,
+                f"share/spack/setup-env.{platform_script_ext}",
+            ),
+        ],
+        ["spack", "gpg", "trust", mounted_gpg_path if job_image else gpg_path] if gpg_path else [],
+        ["spack", "env", "activate", mounted_env_dir if job_image else repro_dir],
+        [
+            (
+                os.path.join(mounted_repro_dir, f"install.{platform_script_ext}")
+                if job_image
+                else install_script
+            )
+        ],
+    ]
+    entry_script = os.path.join(mounted_workdir, f"entrypoint.{platform_script_ext}")
+    inst_list = []
+    # Finally, print out some instructions to reproduce the build
+    if job_image:
+        # Allow interactive
+        install_mechanism = (
+            os.path.join(mounted_repro_dir, f"install.{platform_script_ext}")
+            if job_image
+            else install_script
+        )
+        entrypoint_script.append(["echo", f"Re-run install script using:\n\t{install_mechanism}"])
+        # Allow interactive
+        if IS_WINDOWS:
+            entrypoint_script.append(["&", "($args -Join ' ')", "-NoExit"])
+        else:
+            entrypoint_script.append(["exec", "$@"])
+
+        process_command(
+            "entrypoint", entrypoint_script, work_dir, run=False, exit_on_failure=False
+        )
+
+        docker_command = [
+            runtime,
+            "run",
+            "-i",
+            "-t",
+            "--rm",
+            "--name",
+            "spack_reproducer",
+            "-v",
+            ":".join([work_dir, mounted_workdir, "Z"]),
+            "-v",
+            ":".join(
+                [
+                    os.path.join(work_dir, "jobs_scratch_dir"),
+                    os.path.join(mount_as_dir, "jobs_scratch_dir"),
+                    "Z",
+                ]
+            ),
+            "-v",
+            ":".join([os.path.join(work_dir, "spack"), mount_as_dir, "Z"]),
+            "--entrypoint",
+        ]
+        if IS_WINDOWS:
+            docker_command.extend(["powershell.exe", job_image, entry_script, "powershell.exe"])
+        else:
+            docker_command.extend([entry_script, job_image, "bash"])
```

Comment:
While you are you, can you make this `/bin/sh` instead of `bash`. This may not work in some containers and using `sh` is consistent with how we write `process_command` scripts for not windows.

Reply:
Can it wait til a subsequent PR?  It's the kind of thing I don't feel comfortable changing without repeating all the testing in a real gitlab environment.

Reply:
Yeah, I can make one after this goes in.